frontend/src/components/CoachMark/CoachMark.tsx

```diff
@@ -0,0 +1,308 @@
+    localStorage.removeItem(hideCoachMarkStorageKey);
+  }
+
+  const [isOpen, setIsOpen] = useState(!(isExpired || isDimissed || isDisabled));
+
+  const [coachMark, setCoachMark] = useState<HTMLElement | null>(null);
+  const [referenceElement, setReferenceElement] = useState<HTMLElement | null>(null);
+
+  const [labelId, setLabelId] = useState<string | undefined>();
+  const [descriptionId, setDescriptionId] = useState<string | undefined>();
+
+  const handleDismiss = useCallback(() => {
+    setIsOpen(false);
+    setHideCoachMarkStorageItem(options.id);
+    referenceElement?.classList.remove("coachMarkActive");
+    logEvent("coach_mark_single_feat_dimiss", {
+      coach_mark_id: options.id,
+    });
+  }, [options.id, referenceElement?.classList]);
+
+  useEffect(() => {
+    if (options.referenceRef.current !== null) {
+      setReferenceElement(options.referenceRef.current);
+    }
+  }, [options.referenceRef]);
+
+  const floatingRootContext = useFloatingRootContext({
+    open: isOpen,
+    onOpenChange(nextOpen) {
+      if (isOpen && nextOpen === false) {
+        handleDismiss();
+      } else {
+        setIsOpen(nextOpen);
+      }
+    },
+    elements: {
+      reference: referenceElement,
+      floating: coachMark,
+    },
+  });
+
+  const click = useClick(floatingRootContext, { enabled: false });
+  const dismiss = useDismiss(floatingRootContext, { referencePress: true, outsidePress: false });
+  const interactions = useInteractions([dismiss, click]);
+
+  return useMemo(
+    () => ({
+      referenceElement,
+      setCoachMark,
+      isOpen,
+      handleDismiss,
+      coachMarkPortal: options.coachMarkPortal,
+      coachMarkId: options.id,
+      labelId,
+      descriptionId,
+      setLabelId,
+      setDescriptionId,
+      ...interactions,
+      ...floatingRootContext,
+    }),
+    [
+      referenceElement,
+      isOpen,
+      handleDismiss,
+      options.coachMarkPortal,
+      options.id,
+      labelId,
+      descriptionId,
+      interactions,
+      floatingRootContext,
+    ]
+  );
+};
+
+type CoachMarkContextType = ReturnType<typeof useCoachMark> | null;
+
+const CoachMarkContext = createContext<CoachMarkContextType>(null);
+
+export const useCoachMarkContext = () => {
+  const context = useContext(CoachMarkContext);
+  if (context == null) {
+    throw new Error("CoachMark components must be wrapped in <CoachMark.Root />");
+  }
+
+  return context;
+};
+
+interface CoachMarkRootProps {
+  coachMark: ReturnType<typeof useCoachMark>;
+  children: ReactNode;
+}
+
+const CoachMarkRoot = (props: CoachMarkRootProps) => {
+  return (
+    <CoachMarkContext.Provider value={props.coachMark}>
+      <div data-testid="coach-mark" />
+      {props.coachMark.coachMarkPortal}
+      {props.children}
+    </CoachMarkContext.Provider>
+  );
+};
+
+interface CoachMarkPortalProps {
+  allowedPlacements: Placement[];
+  children: ReactNode;
+}
+
+const CoachMarkPortal = (props: CoachMarkPortalProps) => {
```

Comment:
Agree with the other discussion that this file should be broken up. It's a little hard to tell what's going on with multiple JSX components sharing a file